metomi/rose/loc_handlers/git.py

```diff
@@ -0,0 +1,180 @@
+            if self.git_version < (2, 25, 0):
+                self.manager.popen.run_ok(
+                    "git", git_dir_opt, "config", "extensions.partialClone",
+                    "true"
+                )
+
+            # Extract the commit hash if we don't already have it.
+            commithash = self._get_commithash(remote, ref)
+
+            # Fetch the ref/commit as efficiently as we can.
+            ret_code, _, stderr = self.manager.popen.run(
+                "git", git_dir_opt, "fetch", "--depth=1",
+                "--filter=blob:none", remote, commithash
+            )
+            if ret_code:
+                raise ValueError(f"source={loc.name}: {stderr}")
+
+            # Determine the type of the path object.
+            ret_code, typetext, stderr = self.manager.popen.run(
+                "git", git_dir_opt, "cat-file", "-t", f"{commithash}:{path}"
+            )  # N.B. git versions >1.8 can use '-C' to set git dir.
+            if ret_code:
+                raise ValueError(f"source={loc.name}: {stderr}")
+
+        if typetext.strip() == "tree":
+            loc.loc_type = loc.TYPE_TREE
+        else:
+            loc.loc_type = loc.TYPE_BLOB
+        loc.real_name = (
+            f"remote:{remote} ref:{ref} commit:{commithash} path:{path}"
+        )
+        loc.key = commithash
+
+    async def pull(self, loc, conf_tree):
+        """Get loc to its cache.
+
+        git sparse-checkout is not available below Git 2.25, and seems to omit
+        contents altogether if set to the root of the repo (as of 2.40.1).
+
+        """
+        if not loc.real_name:
+            self.parse(loc, conf_tree)
+        remote, path, ref = self._parse_name(loc)
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            git_dir_opt = f"--git-dir={tmpdirname}/.git"
+            await self.manager.popen.run_ok_async(
+                "git", git_dir_opt, "init"
+            )
+            if self.git_version >= (2, 25, 0) and path != "./":
+                await self.manager.popen.run_ok_async(
+                    "git", git_dir_opt, "sparse-checkout", "set", path,
+                    "--no-cone"
+                )
+                await self.manager.popen.run_ok_async(
+                    "git", git_dir_opt, "fetch", "--depth=1",
+                    "--filter=blob:none", remote, loc.key
+                )
+            else:
+                await self.manager.popen.run_ok_async(
+                    "git", git_dir_opt, "fetch", "--depth=1", remote, loc.key
+                )
+
+            await self.manager.popen.run_ok_async(
+                "git", git_dir_opt, f"--work-tree={tmpdirname}", "checkout",
+                loc.key
+            )
+            name = tmpdirname + "/" + path
+            dest = loc.cache
+            if loc.loc_type == "tree":
+                dest += "/"
+            cmd = self.manager.popen.get_cmd("rsync", name, dest)
+            await self.manager.popen.run_ok_async(*cmd)
```

Comment:
I'm getting a failure here:

```
[INFO] 2024-06-03T11:56:18+0100 rsync -a '--exclude=.*' --timeout=1800 '--rsh=ssh -oBatchMode=yes -oStrictHostKeyChecking=no -oConnectTimeout=8' /var/tmp/tmpzpxb2lpj/README.md /var/tmp/tmpjul8aid7/e99d80506f38bfee39893fd042f2f5d6
[FAIL] 2024-06-03T11:56:19+0100 rsync -a '--exclude=.*' --timeout=1800 '--rsh=ssh -oBatchMode=yes -oStrictHostKeyChecking=no -oConnectTimeout=8' /var/tmp/tmpzpxb2lpj/README.md /var/tmp/tmpjul8aid7/e99d80506f38bfee39893fd042f2f5d6 # return-code=3, stderr=
[FAIL] 2024-06-03T11:56:19+0100 rsync: change_dir#3 "/var/tmp/tmpjul8aid7" failed: No such file or directory (2)
[FAIL] 2024-06-03T11:56:19+0100 rsync error: errors selecting input/output files, dirs (code 3) at main.c(694) [Receiver=3.1.2]
```

Reply:
(status update - we're finding it hard to independently reproduce this one, currently a mystery)